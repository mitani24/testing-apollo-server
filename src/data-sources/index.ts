import { Config } from "apollo-server-express";
import { TrackAPI } from "./track-api";

export type ContextType = {
  dataSources: {
    trackAPI: TrackAPI;
  };
};

export const dataSources: Config["dataSources"] = () => {
  return {
    trackAPI: new TrackAPI(),
  };
};
