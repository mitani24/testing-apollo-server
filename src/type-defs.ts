import fs from "fs";
import path from "path";
import { gql } from "apollo-server";

const schema = fs
  .readFileSync(path.join(__dirname, "../schema.graphql"))
  .toString();

export const typeDefs = gql`
  ${schema}
`;
