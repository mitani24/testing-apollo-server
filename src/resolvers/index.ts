import { ContextType } from "../data-sources";
import { Resolvers } from "../lib/generated/graphql-resolver-types";

export const resolvers: Resolvers<ContextType> = {
  Author: {
    id: ({ id }) => id,
    name: ({ name }) => name,
    photo: ({ photo }) => photo ?? null,
  },
  Mutation: {
    async updateTrack(_, { input }, { dataSources }) {
      const track = await dataSources.trackAPI.updateTrack({
        path: {
          track_id: input.id,
        },
        body: {
          title: input.title ?? undefined,
          description: input.description ?? undefined,
          author_id: input.authorId ?? undefined,
          thumbnail: input.thumbnail ?? undefined,
          length: input.length ?? undefined,
          number_of_views: input.numberOfViews ?? undefined,
        },
      });
      return { track };
    },
  },
  Query: {
    tracks(_, { authorId }, { dataSources }) {
      return dataSources.trackAPI.getTracks({
        query: {
          author_id: authorId ?? undefined,
        },
      });
    },
    track(_, { id }, { dataSources }) {
      return dataSources.trackAPI.getTrack({
        path: {
          track_id: id,
        },
      });
    },
  },
  Track: {
    id: ({ id }) => id,
    title: ({ title }) => title,
    author: ({ author_id }, _, { dataSources }) => {
      return dataSources.trackAPI.getAuthor({
        path: {
          author_id,
        },
      });
    },
    thumbnail: ({ thumbnail }) => thumbnail ?? null,
    length: ({ length }) => length ?? null,
    description: ({ description }) => description ?? null,
    numberOfViews: ({ number_of_views }) => number_of_views ?? null,
  },
  UpdateTrackPayload: {
    track: ({ track }) => track,
  },
};
