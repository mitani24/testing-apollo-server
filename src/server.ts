import { ApolloServer } from "apollo-server";
import { typeDefs } from "./type-defs";
import { dataSources } from "./data-sources";
import { resolvers } from "./resolvers";

export const server = new ApolloServer({
  typeDefs,
  dataSources,
  resolvers,
});
