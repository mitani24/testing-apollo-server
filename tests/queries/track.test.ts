import { HTTPCache } from "apollo-datasource-rest";
import { UserInputError } from "apollo-server";
import { server } from "../../src/server";
import { Track } from "../lib/generated/graphql-documents";

const fetch = jest.spyOn(HTTPCache.prototype, "fetch");

describe("track", () => {
  it("トラックを取得できる", async () => {
    const res = await server.executeOperation({
      query: Track,
      variables: { trackId: "track-1" },
    });

    expect(res).toMatchSnapshot();
  });

  it("/tracks/<track_id> エンドポイントを叩く", async () => {
    const res = await server.executeOperation({
      query: Track,
      variables: { trackId: "track-1" },
    });

    expect(fetch).toHaveBeenCalledTimes(2);
    const reqs = fetch.mock.calls.map(([req]) => req);
    const urls = reqs.map((req) => new URL(req.url));
    expect(urls[0].pathname).toContain("/tracks/track-1");
  });

  it("track_id 未指定だとエラーが返る", async () => {
    const res = await server.executeOperation({
      query: Track,
    });

    expect(res.errors).toHaveLength(1);
    expect(res.errors[0]).toBeInstanceOf(UserInputError);
  });
});
