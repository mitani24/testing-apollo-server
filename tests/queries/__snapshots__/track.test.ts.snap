// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`track トラックを取得できる 1`] = `
Object {
  "data": Object {
    "track": Object {
      "author": Object {
        "id": "string",
        "name": "string",
        "photo": "string",
      },
      "description": "string",
      "id": "string",
      "length": 0,
      "numberOfViews": 0,
      "thumbnail": "string",
      "title": "string",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
